### PR TITLE
Display origin auth failure in a dialog

### PR DESCRIPTION
**Pull request by ASpoonPlaysGames**
Currently, we don't do anything with origin auth failure, meaning the next request just fails, and we get not particularly relevant error messages (player not found, invalid masterserver token)

Depends on
- https://github.com/R2Northstar/NorthstarLauncher/pull/468

Still todo:
- [x] Localisation
- [x] Cooler links

### CODE REVIEW

**Comment by F1F7Y** [comment]
>     * [ ]  Localisation

Do you plan on adding other languages?

**Comment by ASpoonPlaysGames** [comment]
> Do you plan on adding other languages?

I would like to tbh, but I'm not rly fussed at this point

**Comment by GeckoEidechse** [comment]
> > Do you plan on adding other languages?
> 
> I would like to tbh, but I'm not rly fussed at this point

There's no need to add other languages in original PR. We got a [weblate setup](https://translate.harmony.tf/projects/northstar/client/) via [translate.harmony.tf](https://translate.harmony.tf) where it will show any untranslated strings and translators can add them via weblate once the PR has been merged.

**Comment by Alystrasz** [comment]
> translators can add them via weblate once the PR has been merged.

We gotta be careful about merging in *all* translations before releasing, then.

**Comment by GeckoEidechse** [comment]
> > translators can add them via weblate once the PR has been merged.
> 
> We gotta be careful about merging in _all_ translations before releasing, then.

Eh, like we don't even have everything translated yet, so if we do that we would just wait forever

image
https://translate.harmony.tf/projects/northstar/client/

**Comment by Alystrasz** [comment]
> Eh, like we don't even have everything translated yet, so if we do that we would just wait forever

French is now 100% :smile: 
Gotta ping people in the appropriate Discord channel.

File: Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	if ( res.errorCode != "" )
+		dialogData.message += "\n\n" + Localize( "#AUTHENTICATION_FAILED_ERROR_CODE", res.errorCode )
+
+
+	CloseAllDialogs()
+	AddDialogButton( dialogData, "#OK", null )
+	AddDialogButton( dialogData, Localize( "#AUTHENTICATION_FAILED_HELP" ), void function() {
+		// todo: get MS to redirect, so i can use an MS link or something?
+		// also todo: link using # to the error code given, needs work on wiki side of things
+		string link = format( "%s/wiki", GetConVarString( "ns_masterserver_hostname" ) )
```
Comment (NoCatt) [inline Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut]:
```suggestion
		string link = "https://r2northstar.gitbook.io/r2northstar-wiki/installing-northstar/troubleshooting"
```
we can change this to ``format( "https://r2northstar.gitbook.io/r2northstar-wiki/installing-northstar/troubleshooting#%s, GetErrorLink( error ) `` in the future

File: Northstar.Client/mod/resource/northstar_client_localisation_english.txt [hunk 2]
```diff
@@ -321,6 +321,11 @@ Press Yes if you agree to this. This choice can be changed in the mods menu at a
 		"INVALID_MASTERSERVER_TOKEN" 	"Invalid or expired masterserver token"
 		"JSON_PARSE_ERROR" 				"Error parsing json response"
 		"UNSUPPORTED_VERSION" 			"The version you are using is no longer supported"
+
+		"AUTHENTICATION_FAILED_HEADER"		"Authentication Failed"
+		"AUTHENTICATION_FAILED_BODY"		"Failed to authenticate with Atlas!"
```
Comment (NoCatt) [inline Northstar.Client/mod/resource/northstar_client_localisation_english.txt:326]: no one who cant fix it themself knows what atlas is, I feel like Masterserver is more understanable

**Review by NoCatt — CHANGES_REQUESTED** [review]

File: Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut (hunk 1 above)
Comment (NoCatt) [reply Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut]: I couldnt remove it we had this in the discord already just mark as resolved

File: Northstar.Client/mod/resource/northstar_client_localisation_english.txt (hunk 2 above)
Comment (ASpoonPlaysGames) [reply Northstar.Client/mod/resource/northstar_client_localisation_english.txt:326]: maaaaaan, we really need to start moving away from calling it the masterserver when it is called Atlas now

Comment (NoCatt) [reply Northstar.Client/mod/resource/northstar_client_localisation_english.txt:326]: its not for me okay, its for the northstar chat users, lets hear a 3rd party on this

**Comment by ASpoonPlaysGames** [comment]
image

here is an example of how the dialog looks in game

**Review by NoCatt — APPROVED** [review]
Colour and layout looks good, the help will now take the common errors to the right wiki section. Code is kept to a minimum ( thats good ).

**Comment by ASpoonPlaysGames** [comment]
I've also tested this pretty thoroughly when making the colours and designing the UI. But if someone else can test this that would be great. Best way I found is to change your origin token with a debugger to get a bad response back from stryder when you try and auth

File: Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	string link = "https://r2northstar.gitbook.io/r2northstar-wiki/installing-northstar/troubleshooting"
+	// link to generic troubleshooting page if we don't have an error code from Atlas
+	if ( res.errorCode != "" )
+		link = format( "%s#%s", link, res.errorCode )
+
+	CloseAllDialogs()
+	AddDialogButton( dialogData, "#OK" )
+	AddDialogButton( dialogData, Localize( "#AUTHENTICATION_FAILED_HELP" ), void function() : ( dialogData, link )
+	{
+		// todo: get MS to redirect, so i can use an MS link or something?
```
Comment (Alystrasz) [inline Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut:49]: Still TODO?

**Review by Alystrasz — CHANGES_REQUESTED** [review]
Information exposed in https://github.com/R2Northstar/NorthstarLauncher/pull/468 is correctly displayed in a dialog.
(test done by manually corrupting the [`uidStr` variable](https://github.com/R2Northstar/NorthstarLauncher/blob/main/NorthstarDLL/masterserver/masterserver.cpp#L95) through the debugger)

image

File: Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut (hunk 3 above)
Comment (ASpoonPlaysGames) [reply Northstar.Client/mod/scripts/vscripts/ui/atlas_auth.nut:49]: Yes and no, MS *does* redirect from `https://northstar.tf/wiki` but any child page of that does not redirect. This means that I can't link to `https://northstar.tf/wiki/installing-northstar/troubleshooting` or anything like that

**Review by NoCatt — APPROVED** [review]
code looks good, the link opened automatically jumps to the correct position in the troubleshooting section ( I made a PR for that )
Otherwise this will help with tickets and finding the issue, and hopefully prevent some

**Comment by GeckoEidechse** [comment]
Merging based on reviewer feedback. Did not really properly look at this myself

**Review by GeckoEidechse — APPROVED** [review]
Signing off based on reviewer feedback. Did not really properly look at this myself.

**Pull request merged**